python_modules/dagster/dagster/_daemon/asset_daemon.py

```diff
@@ -916,15 +963,35 @@ def _evaluate_auto_materialize_tick(
             # Write out the persistent cursor, which ensures that future ticks will move on once
             # they determine that nothing needs to be retried
             if sensor:
-                state = instance.get_instigator_state(
-                    sensor.get_external_origin_id(), sensor.selector_id
+                evaluation_states = new_cursor.previous_evaluation_state_by_key
+                instance.daemon_cursor_storage.set_cursor_values(
+                    {
+                        _asset_key_to_cursor_key(asset_key): serialize_value(
+                            AssetScopedAssetDaemonCursor(state)
+                        )
+                        for asset_key, state in evaluation_states.items()
+                    }
+                )
+
+                # Eval states are sourced from individual AssetScopedAssetDaemonCursors, so we don't save them
+                # on this object.
+                new_cursor_without_eval_state = dataclasses.replace(
+                    new_cursor, previous_evaluation_state=[]
+                )
+
+                state = check.not_none(
+                    instance.get_instigator_state(
+                        sensor.get_external_origin_id(), sensor.selector_id
+                    )
                 )
                 instance.update_instigator_state(
```

Comment:
what happens if this first daemon_cursor_storage write succeeds and the second update_instigator_state write fails? Does that leave things in a broken state on the next tick? We could try to do it in a transaction so that that is not possible (might be a pain since they are in different storage classes - daemon_cursor_storage is in run storage but the instigator_state stuff is in schedule storage)